include/nbl/builtin/hlsl/sort/counting.hlsl

```diff
@@ -0,0 +1,123 @@
+    {
+        nbl::hlsl::glsl::barrier();
+    }
+};
+
+static ScratchProxy arithmeticAccessor;
+
+groupshared uint32_t sdata[BucketCount];
+
+template<typename KeyAccessor, typename ValueAccessor, typename ScratchAccessor>
+struct counting
+{
+    void histogram(NBL_REF_ARG(KeyAccessor) key, NBL_REF_ARG(ScratchAccessor) scratch, const nbl::hlsl::sort::CountingPushData data)
+    {
+        uint32_t tid = nbl::hlsl::workgroup::SubgroupContiguousIndex();
+
+        [unroll]
+        for (int i = 0; i < BucketsPerThread; i++)
+            sdata[BucketsPerThread * tid + i] = 0;
+        uint32_t index = (nbl::hlsl::glsl::gl_WorkGroupID().x * WorkgroupSize) * data.elementsPerWT;
+
+        nbl::hlsl::glsl::barrier();
+
+        for (int i = 0; i < data.elementsPerWT; i++)
+        {
+            int j = index + i * WorkgroupSize + tid;
+            if (j >= data.dataElementCount)
+                break;
+            uint32_t k = key.get(j);
+            nbl::hlsl::glsl::atomicAdd(sdata[k - data.minimum], (uint32_t) 1);
+        }
+
+        nbl::hlsl::glsl::barrier();
+
+        uint32_t sum = 0;
+        uint32_t scan_sum = 0;
+
+        for (int i = 0; i < BucketsPerThread; i++)
+        {
+            sum = nbl::hlsl::workgroup::exclusive_scan < nbl::hlsl::plus < uint32_t >, WorkgroupSize > ::
+            template __call <ScratchProxy>
+            (sdata[WorkgroupSize * i + tid], arithmeticAccessor);
+
+            arithmeticAccessor.workgroupExecutionAndMemoryBarrier();
+
+            nbl::hlsl::glsl::atomicAdd(scratch.get_ptr(WorkgroupSize * i + tid), sum);
+            if ((tid == WorkgroupSize - 1) && i > 0)
+                nbl::hlsl::glsl::atomicAdd(scratch.get_ptr(WorkgroupSize * i), scan_sum);
+
+            arithmeticAccessor.workgroupExecutionAndMemoryBarrier();
+
+            if ((tid == WorkgroupSize - 1) && i < (BucketsPerThread - 1))
+            {
+                scan_sum = sum + sdata[WorkgroupSize * i + tid];
+                sdata[WorkgroupSize * (i + 1)] += scan_sum;
+            }
+        }
+    }
+                
+    void scatter(NBL_REF_ARG(KeyAccessor) key, NBL_REF_ARG(ValueAccessor) val, NBL_REF_ARG(ScratchAccessor) scratch, const nbl::hlsl::sort::CountingPushData data)
+    {
+        uint32_t tid = nbl::hlsl::workgroup::SubgroupContiguousIndex();
+
+        [unroll]
+        for (int i = 0; i < BucketsPerThread; i++)
+            sdata[BucketsPerThread * tid + i] = 0;
+        uint32_t index = (nbl::hlsl::glsl::gl_WorkGroupID().x * WorkgroupSize) * data.elementsPerWT;
+
+        nbl::hlsl::glsl::barrier();
+
+        [unroll]
+        for (int i = 0; i < data.elementsPerWT; i++)
+        {
+            int j = index + i * WorkgroupSize + tid;
+            if (j >= data.dataElementCount)
+                break;
+            uint32_t k = key.get(j);
+            uint32_t v = val.get(j);
+            sdata[k - data.minimum] = nbl::hlsl::glsl::atomicAdd(scratch.get_ptr(k - data.minimum), (uint32_t) 1);
```

Comment:
what you can do is "unmerge" (subtract local non-cumulative histogram from global cumulative) the histograms via an atomic subtraction, then store the return value minus the atomic sub operand to `scratch` (shared memory)

Then you'll have a start index to place your elements into, which you can atomicAdd(,1) to get your actual global offset.

Then in your final loop where you actually look at the keys again, and work out their global placement position 

Pseudocode
```
initialize scratch to 0
barrier
histogram
barrier
unmerge histogram
barrier
pseudo histogram & scatter
```

Reply:
the way you'd unmerge is
```cpp
sdata[index] = histogram.atomicSub(index,sdata[index])-sdata[index];
```
now you have the lowest address an element with `key-minimum==index` would get written into

Reply:
and what I mean by pseudohistogram & scatter is
```cpp
const Key k = key.get(j);
if (robust...)
  cotninue;
const sortedIx = scratch.atomicAdd(k-data.minimum,1);
key.set(sortedIx,k);
```

I leave the moving of the Value as an exercise to the reader.

Reply:
Done